The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6834855318/job/18588011556) [(plain)](https://github.com/rust-lang/rust/commit/f95183e418a35972bb94255142a6b0390007ed3e/checks/18588011556/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_99931284-9a9a-459f-b390-d6a2ebcd7d95
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=properly-reject-defaultness-on-free-consts
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_99931284-9a9a-459f-b390-d6a2ebcd7d95
GITHUB_REF=refs/pull/117818/merge
GITHUB_REF_NAME=117818/merge
GITHUB_REF_PROTECTED=false
---
25 error: `default` is only allowed on items in trait impls
-   --> $DIR/trait-item-with-defaultness-fail-semantic.rs:9:5
+   --> $DIR/defaultness-invalid-places-fail-semantic.rs:9:5
27    |
28 LL |     default type C: Ord;

31    |     `default` because of this
32 
33 error: `default` is only allowed on items in trait impls
---
--- stderr -------------------------------
error: `default` is only allowed on items in trait impls
##[error]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:6:5
   |
LL |     default const A: u8; //~ ERROR `default` is only allowed on items in trait impls
   |     |
   |     `default` because of this

error: `default` is only allowed on items in trait impls
error: `default` is only allowed on items in trait impls
##[error]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:7:5
   |
LL |     default const B: u8 = 0;  //~ ERROR `default` is only allowed on items in trait impls
   |     |
   |     `default` because of this

error: `default` is only allowed on items in trait impls
error: `default` is only allowed on items in trait impls
##[error]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:8:5
   |
LL |     default type D; //~ ERROR `default` is only allowed on items in trait impls
   |     |
   |     `default` because of this

error: `default` is only allowed on items in trait impls
error: `default` is only allowed on items in trait impls
##[error]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:9:5
   |
LL |     default type C: Ord; //~ ERROR `default` is only allowed on items in trait impls
   |     |
   |     `default` because of this

error: `default` is only allowed on items in trait impls
error: `default` is only allowed on items in trait impls
##[error]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:10:5
   |
LL |     default fn f1(); //~ ERROR `default` is only allowed on items in trait impls
   |     |
   |     `default` because of this

error: `default` is only allowed on items in trait impls
error: `default` is only allowed on items in trait impls
##[error]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:11:5
   |
LL |     default fn f2() {} //~ ERROR `default` is only allowed on items in trait impls
   |     |
   |     `default` because of this

warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs:1:12
   |
LL | #![feature(specialization)] //~ WARN the feature `specialization` is incomplete
   |
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: consider using `min_specialization` instead, which is more stable and complete
   = note: `#[warn(incomplete_features)]` on by default
```

</details>

--- a/compiler/rustc_ast_passes/src/ast_validation.rs
+++ b/compiler/rustc_ast_passes/src/ast_validation.rs
@@ -1008,12 +1008,14 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                     _ => {}
                 }
             }
-            ItemKind::Const(box ConstItem { defaultness, expr: None, .. }) => {
+            ItemKind::Const(box ConstItem { defaultness, expr, .. }) => {
                 self.check_defaultness(item.span, *defaultness);
-                self.session.emit_err(errors::ConstWithoutBody {
-                    span: item.span,
-                    replace_span: self.ending_semi_or_hi(item.span),
-                });
+                if expr.is_none() {
+                    self.session.emit_err(errors::ConstWithoutBody {
+                        span: item.span,
+                        replace_span: self.ending_semi_or_hi(item.span),
+                    });
+                }
             }
             ItemKind::Static(box StaticItem { expr: None, .. }) => {
                 self.session.emit_err(errors::StaticWithoutBody {

--- a/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs
+++ b/tests/ui/parser/defaultness-invalid-places-fail-semantic.rs
@@ -10,3 +10,7 @@ trait X {
     default fn f1(); //~ ERROR `default` is only allowed on items in trait impls
     default fn f2() {} //~ ERROR `default` is only allowed on items in trait impls
 }
+
+default const E: u8 = 0; //~ ERROR `default` is only allowed on items in trait impls
+default type F = (); //~ ERROR `default` is only allowed on items in trait impls
+default fn h() {} //~ ERROR `default` is only allowed on items in trait impls

--- a/tests/ui/parser/defaultness-invalid-places-fail-semantic.stderr
+++ b/tests/ui/parser/defaultness-invalid-places-fail-semantic.stderr
@@ -1,5 +1,5 @@
 error: `default` is only allowed on items in trait impls
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:6:5
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:6:5
    |
 LL |     default const A: u8;
    |     -------^^^^^^^^^^^^^
@@ -7,7 +7,7 @@ LL |     default const A: u8;
    |     `default` because of this
 
 error: `default` is only allowed on items in trait impls
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:7:5
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:7:5
    |
 LL |     default const B: u8 = 0;
    |     -------^^^^^^^^^^^^^^^^^
@@ -15,7 +15,7 @@ LL |     default const B: u8 = 0;
    |     `default` because of this
 
 error: `default` is only allowed on items in trait impls
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:8:5
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:8:5
    |
 LL |     default type D;
    |     -------^^^^^^^^
@@ -23,7 +23,7 @@ LL |     default type D;
    |     `default` because of this
 
 error: `default` is only allowed on items in trait impls
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:9:5
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:9:5
    |
 LL |     default type C: Ord;
    |     -------^^^^^^^^^^^^^
@@ -31,7 +31,7 @@ LL |     default type C: Ord;
    |     `default` because of this
 
 error: `default` is only allowed on items in trait impls
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:10:5
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:10:5
    |
 LL |     default fn f1();
    |     -------^^^^^^^^^
@@ -39,15 +39,39 @@ LL |     default fn f1();
    |     `default` because of this
 
 error: `default` is only allowed on items in trait impls
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:11:5
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:11:5
    |
 LL |     default fn f2() {}
    |     -------^^^^^^^^
    |     |
    |     `default` because of this
 
+error: `default` is only allowed on items in trait impls
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:14:1
+   |
+LL | default const E: u8 = 0;
+   | -------^^^^^^^^^^^^^^^^^
+   | |
+   | `default` because of this
+
+error: `default` is only allowed on items in trait impls
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:15:1
+   |
+LL | default type F = ();
+   | -------^^^^^^^^^^^^^
+   | |
+   | `default` because of this
+
+error: `default` is only allowed on items in trait impls
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:16:1
+   |
+LL | default fn h() {}
+   | -------^^^^^^^
+   | |
+   | `default` because of this
+
 warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
-  --> $DIR/trait-item-with-defaultness-fail-semantic.rs:1:12
+  --> $DIR/defaultness-invalid-places-fail-semantic.rs:1:12
    |
 LL | #![feature(specialization)]
    |            ^^^^^^^^^^^^^^
@@ -56,5 +80,5 @@ LL | #![feature(specialization)]
    = help: consider using `min_specialization` instead, which is more stable and complete
    = note: `#[warn(incomplete_features)]` on by default
 
-error: aborting due to 6 previous errors; 1 warning emitted
+error: aborting due to 9 previous errors; 1 warning emitted
 

